### PR TITLE
Implement `Element` for `bool`

### PR DESCRIPTION
**Pull request by laggui**
### Checklist

- [x] Confirmed that `run-checks all` script has been executed.

### Related Issues/PRs

Required for #1535 and `Data` refactor

### Changes

Since the traits `Zero`, `One` and `ToPrimitive` from `num-traits` could not be implemented for `bool`:
- ~~Added new `Zero` and `One` element identities (without `Add` and `Mul` requirement)~~ Removed the traits (previously kept for historical reasons)
- Added new `ToElement` trait (adapted from `num_traits::ToPrimitive`)

`Element` is now implemented for `bool`, which should simplify a couple of things.

### CODE REVIEW

File: crates/burn-ndarray/src/element.rs [hunk 1]
```diff
@@ -20,7 +19,6 @@ where
 /// A general element for ndarray backend.
 pub trait NdArrayElement:
     Element
-    + One
```
Comment (laggui) [inline crates/burn-ndarray/src/element.rs:23]: `Element` already requires `One`

File: crates/burn-ndarray/src/ops/base.rs [hunk 2]
```diff
@@ -567,12 +567,14 @@ where
             tensor
                 .array
                 .mapv(|x| {
-                    if x > E::zero() {
-                        E::one()
-                    } else if x < E::zero() {
-                        -E::one()
+                    let zero = 0.elem();
+                    let one = 1.elem::<E>();
```
Comment (nathanielsimard) [inline crates/burn-ndarray/src/ops/base.rs:571]: Do you think it would be better to declare `zero` and `one` outside the `mapv` ?

**Comment by laggui** [comment]
Not sure why clippy failed on that commit specifically, seems unrelated.

**Review by antimora — APPROVED** [review]
LGTM

**Comment by antimora** [comment]
Will this fix #1535?

**Comment by laggui** [comment]
> Will this fix #1535?

The ops won't be implemented in this PR but with these changes they could be refactored properly.

**Review by nathanielsimard — APPROVED** [review]
LGTM, however I would add num_trait to the NOTICE.md file before merging.

**Pull request merged**